Optimization of nearest location transfer for nearest-subapp

GiudGiud opened the pull request:
the nearest-position mode (which could cover nearest-subapp) was taking a big hit on performance on cases with very many subapps, because it was creating a KDtree for every (global) sub-app
This fixes that problem by adding the nearest-app capability in its own right

This PR includes:
- an automatic mode for the grid partitioner
- an automatic mode for the hierarchic grid partitioner
- additional exceptions testing for both
- a nearest-app mode for the "which process should I talk to" in general field transfers
- a reworked nearest-app mode for the nearest-node transfer (the user-preferred extrapolation transfer)
- various indexing fixes and more comments and explanations on indices in the code
- a new test and fixes for search_value_conflicts + nearest-positions
- KDTree for retrieving nearest positions (to handle large numbers of positions)
- cleanup of initialize / finalize across all positions so they can be trusted to be finalized when used (even when Problem/solve=false)

GiudGiud committed:
Optimize nearest position retrieval
refs #23587

GiudGiud committed:
Add support for nearest-application mode in NL transfer:
- only use the minimal number of KDTree (# local apps) rather than the number of positions (global #of apps)
refs #23585

Apply suggestions from code review

GiudGiud committed:
Improve nearest-position check by only comparing index
Add a special case bypassing bounding boxes for nearest-app mode to determine the source processors
refs #28782

GiudGiud commented on framework/src/positions/Positions.C:
```diff
@@ -50,6 +50,12 @@ Positions::Positions(const InputParameters & parameters)
+  if (_initial_positions)
+  {
+    auto initial_positions_copy = *_initial_positions;
+    _initial_positions_kd_tree = std::make_unique<KDTree>(initial_positions_copy, 1);
```
this is probably a terrible idea.
why does the KDTree not accept a const reference? I ll try to fix that

GiudGiud replied:
done

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -90,6 +94,15 @@ MultiAppGeneralFieldNearestLocationTransfer::MultiAppGeneralFieldNearestLocation
+        "When using the 'nearest child application' data, the source data (positions and values) "
+        "are grouped on a per-application basis, so it cannot be agglomerated over all child "
+        "applications.\nNote that the option to use nearest applications for source restrictions, "
+        "but further split each child application's domain in there regions closest to each child "
```
```suggestion
        "but further split each child application's domain by regions closest to each position (here the the child "
```

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -90,6 +94,15 @@ MultiAppGeneralFieldNearestLocationTransfer::MultiAppGeneralFieldNearestLocation
+        "are grouped on a per-application basis, so it cannot be agglomerated over all child "
+        "applications.\nNote that the option to use nearest applications for source restrictions, "
+        "but further split each child application's domain in there regions closest to each child "
+        "application's centroid (which could be conceived when 'group_subapps' = false) is also "
```
```suggestion
        "application's centroid), which could be conceived when 'group_subapps' = false, is also "
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2483355/), step Clang format on 01050a6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9fb4f87899b841a2a4a440d625f842b4809d7ed0`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2483406/), step Clang format on d7f771f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9fb4f87899b841a2a4a440d625f842b4809d7ed0`

GiudGiud committed:
Avoid copying initial positions reporter vector

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2483458/), step Clang format on f675883 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9fb4f87899b841a2a4a440d625f842b4809d7ed0`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2483517/), step Clang format on a42f71a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9fb4f87899b841a2a4a440d625f842b4809d7ed0`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2483601/), step Docs: sync website on d961b8d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28785/site)

This comment will be updated on new commits.

GiudGiud committed:
Add an automatic, but possibly constrained method for the GridPartitioner
Avoid ignoring user parameters when they dont match the number of MPI
closes #28787

GiudGiud committed:
Make hierarchical grid partitioner behavior similar to grid partitioner
- add an automatic option, for each level
- make sure we can constrain the selections

loganharbour commented on framework/src/partitioner/NearestPositionsPartitioner.C:
```diff
@@ -0,0 +1,218 @@
+    positions_array[int(float(i) / n_positions * n_bins)].push_back(positions[i]);
+
+  if (n_bins > n_positions)
+    mooseError("Not implemented for more MPI processes than positions");
```
probably a good use for the documented error?

GiudGiud replied:
Removing this partitioner from the PR. The method worked fine up to a few dozen processes but completely broke after so it's not worth checking

GiudGiud committed:
Add tests for grid partitioner automatic mode + exceptions

GiudGiud committed:
Add tests for hierarchical grid partitioner automated modes + exceptions
refs #28787

GiudGiud committed:
Fix quirks in positions behavior at initial
and for multiapp postions not initializing

GiudGiud committed:
Add testing for nearest-app refactoring

GiudGiud committed:
Fix an issue where shape evaluation transfer was not saving value conflicts in reference frame when using nearest-app/positions
also:
- create multiapp positions with factory
- check distance to positions when comparing nearest target position and nearest app position to handle equidistant case (find the value conflicts)
- accept infinite values if coming from the nearest-app

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2585888/), step Clang format on 96cd824 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b4e638f2f3d244cd8db128f7ea29a3ed135f7c8e`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2585888/), step Fixup commit check on 96cd824 wanted to post the following:

The following commit(s) contain fixups:

- 7428716cc31: fixup! Make hierarchical grid partitioner behavior similar to grid partitioner - add an automatic option, for each level - make sure we can constrain the selections
- 35bdfc3bb73: fixup! Add tests for grid partitioner automatic mode + exceptions
- ddfb6d907e2: fixup! Add an automatic, but possibly constrained method for the GridPartitioner Avoid ignoring user parameters when they dont match the number of MPI closes #28787

GiudGiud committed:
Separate test for value-conflict + nearest app in nodal & elemental

GiudGiud committed:
Modify test to show between multiapps test agrees between
> nearest-app mode and nearest-position mode
> Thus validating the better search heuristic in nearest-ap

GiudGiud committed:
Fix indexing issue in nearest app:
use the global app (=the position index) when checking if the point is nearest to an app

GiudGiud committed:
Fix indexing for nearest-app communication restriction

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2588223/), step Clang format on ac84ce5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 82cdfd96d6d342593d68cc0b089366cf87d0b77c`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2588223/), step Fixup commit check on ac84ce5 wanted to post the following:

The following commit(s) contain fixups:

- 7428716cc31: fixup! Make hierarchical grid partitioner behavior similar to grid partitioner - add an automatic option, for each level - make sure we can constrain the selections
- 35bdfc3bb73: fixup! Add tests for grid partitioner automatic mode + exceptions
- ddfb6d907e2: fixup! Add an automatic, but possibly constrained method for the GridPartitioner Avoid ignoring user parameters when they dont match the number of MPI closes #28787

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2589202/), step Generate coverage on 8dc8cfd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>82cdfd</th>
      <th colspan=4>#28785 8dc8cf</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.18%</td>
      <td>85.17%</td>
      <td>-0.00%</td>
      <td>84.46%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108182</td>
      <td>108248</td>
      <td>+66</td>
      <td>163</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18828</td>
      <td>18843</td>
      <td>+15</td>
      <td>30</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28785/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28785/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28785/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 84.46% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
should be fine now. automatic partitioning is hard

lindsayad commented:
I'll review when more recipes are passing

GiudGiud commented:
I must have removed a commit by accident, these failures are kind of trivial. I ll fix them

GiudGiud committed:
Tune the grid partitioner warning message on ill-specified grid more

GiudGiud commented:
This should be fairly ready? @licharlot tried it on a reactor simulation.
I used it quite a bit too when making the Transfers presentation

lindsayad commented on framework/src/partitioner/GridPartitioner.C:
```diff
@@ -66,22 +75,65 @@ GridPartitioner::_do_partition(MeshBase & mesh, const unsigned int /*n*/)
+                   "will be ignored.");
+    }
+    // 0 means no restriction on which number to choose
+    int dims[] = {isParamValid("nx") ? int(nx) : 0,
```
```suggestion
    std::array<int, 3> dims = {isParamValid("nx") ? int(nx) : 0,
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -277,7 +289,11 @@ MultiAppGeneralFieldNearestLocationTransfer::buildKDTrees(const unsigned int var
           // position to the target point is queried
-          if (_nearest_positions_obj && !closestToPosition(i_pos, transformed_node))
+          // We do not need to check the positions when using nearest app as we will assume
+          // (somewhat incorrectly) that all the points in each subapp are closest to that subapp
```
```suggestion
          // (somewhat incorrectly) that all the points in each subapp are closer to that subapp
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -329,7 +345,10 @@ MultiAppGeneralFieldNearestLocationTransfer::buildKDTrees(const unsigned int var
 
-          if (_nearest_positions_obj && !closestToPosition(i_pos, transformed_vertex_average))
+          // We do not need to check the positions when using nearest app as we will assume
+          // (somewhat incorrectly) that all the points in each subapp are closest to that subapp
```
```suggestion
          // (somewhat incorrectly) that all the points in each subapp are closer to that subapp
```

GiudGiud committed:
Address Alex's review

- better comments

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2640605/), step Clang format on 1e701eb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b4516416a45e370aa097d9e3b4b4aac4cc17e2d3`

GiudGiud commented on framework/src/partitioner/GridPartitioner.C:
```diff
@@ -66,22 +75,65 @@ GridPartitioner::_do_partition(MeshBase & mesh, const unsigned int /*n*/)
+    std::array<int, 3> dims = {isParamValid("nx") ? int(nx) : 0,
+                               isParamValid("ny") ? int(ny) : 0,
+                               isParamValid("nz") ? int(nz) : 0};
+
```
this is failing with
```
/tmp/build/moose/framework/src/partitioner/GridPartitioner.C: In member function 'virtual void GridPartitioner::_do_partition(libMesh::MeshBase&, unsigned int)':
/tmp/build/moose/framework/src/partitioner/GridPartitioner.C:137:47: error: cannot convert 'std::array<int, 3>' to 'int*'
  137 |     MPI_Dims_create(mesh.n_partitions(), dim, dims);
      |                                               ^~~~
      |                                               |
      |                                               std::array<int, 3>
In file included from /opt/libmesh/include/libmesh/libmesh_common.h:61,
```
I ll go back to the classic int []

lindsayad replied:
Nah, use https://en.cppreference.com/w/cpp/container/array/data instead

GiudGiud replied:
Ok I ll get that in a future PR

GiudGiud commented:
Test failures are 3 unrelated timeouts

GiudGiud commented:
Thanks for the review!

roystgnr commented:
@GiudGiud, could this PR have anything to do with the failure here: https://civet.inl.gov/job/2712966/ ?

GiudGiud commented:
the test behaves as expected !
```
    [wrong_numbers]
      type = 'RunException'
      input = 'hierarchical_grid_partitioner.i'
      expect_err = "Partitioning creates 16 partitions, which does not add up to the total number of processors: 8"
      cli_args = 'Mesh/Partitioner/number_nodes=4 Mesh/Partitioner/number_procs_per_node=4 Mesh/Partitioner/processors_grid_computation=automatic Mesh/Partitioner/nodes_grid_computation=automatic Mesh/Partitioner/nx=12'
      min_parallel = 8
      max_parallel = 8
      detail = 'the user specified partitioning parameters for more partitions than there are processes.'
    []
```
this is a testing suite issue. it thinks the output is null


maybe we should only error on process 0 ?

GiudGiud commented:
looks at this beautiful rank 0 output
```
partitioners/hierarchical_grid_partitioner.exceptions/wrong_numbers: Begin redirected output /opt/civet0/build_0/copied_tests/moose_test/tests/partitioners/hierarchical_grid_partitioner/wrong_numbers.processor.0
partitioners/hierarchical_grid_partitioner.exceptions/wrong_numbers: ################################################################################
partitioners/hierarchical_grid_partitioner.exceptions/wrong_numbers: NULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULLNULL

```

roystgnr commented:
`ALL PARTITIONS AND NO PROCESSORS MAKE MOOSE A DULL BOY`

roystgnr commented:
Is there some kind of race condition here?  Rank N errors out fast enough to kill rank 0 before rank 0 can emit the right text to `expect_err`?

For runtime errors I do like to do a parallel communication then emit the error only to `libMesh::err` on processor 0 (to avoid duplicate messages), but MPI likes to kill every process as soon as it sees an abort on one process, and I've never thought about the possibility of race conditions before.

GiudGiud commented:
I guess.

I ll put up a commit in my other transfers PR

lindsayad commented:
It's weird that we're suddenly seeing this on the libmesh PR

GiudGiud commented:
are you seeing it consistently?

lindsayad commented:
we're seeing it consistently for partitioner *tests*, but not always the same test: 

- https://civet.inl.gov/job/2712912/: `partitioners/hierarchical_grid_partitioner.exceptions/wrong_manual_procs_partitioning`
- https://civet.inl.gov/job/2712966/: `partitioners/hierarchical_grid_partitioner.exceptions/wrong_numbers`

GiudGiud commented:
ok my comit should fix it
